Solarer/split culling mode

Solarer opened the pull request:
This is the second attempt to split culling in a mode that is movement restricted to selection and one mode that is not restricted.
This is a prerequisite to get #6025 done.

I tried this before in PR #11112 but the code was rejected. I will leave that PR open for reference until this one has been merged.

Changes:
* implemented a button that is only visible in culling mode and shows up on the right side of the current zoom level. If this button is enabled, the movement is restricted to selected images. If the button is disabled, it is possible to move freely. (This is no new behavior, it is just a visualization of the current culling behavior and allows a user to switch between the two)
* Keybind `c` enters culling but with movement restriction DISABLED (original keybind `x` enters culling and restricts movement - same as current behavior). You can also switch between `c` and `x` at any time to enable/disable movement restriction.

Issues: 
* 'nice' icon for the button is missing
* Layout buttons shift a little left and right because of center alignment when restriction button appears/disappears.

Solarer commented:
@AlicVB it has been some time but I found some time to code recently. Can you have a look at this?

Solarer commented:
@AlicVB
just a friendly ping to ask for your review.

AlicVB commented:
OK, I've done a first try, after a looonng wait (sorry for that). I fear that I will ask you some questions already answered before, but well...

note that this has needed some rebasing (that's quite normal due to the age of the PR)
also there's lot of gtk-critical warnings, certainly related to the new button.

imho, we have 1 main issue here : 
There's no more "automatic" startup switch between restricted and unrestricted mode. That means that users need to remember 2 shortcuts (and use theme accordingly) for the same feature, depending if a selection has been done or not... not really good imho. I wonder why you need to remove this feature ? Even in the case of a "selection-first" act-on algo, you may want to keep the "automatic" startup mode, no ?
If you absolutely need to have a way to enter non-automatic modes, maybe we can have something like : 
- c = unrestricted
- ctrl-c = restricted
- x = automatic

I fear that I still fail to understand the reason behind this change... (what is blocking exactly for your new act-on algo ?)

AlicVB commented:
To go further with my difficulty to understand, I've read again PR #11112 where you say : 
> The changes in this PR are a prerequisit to everything in https://github.com/darktable-org/darktable/pull/10728 and https://github.com/darktable-org/darktable/issues/6025 because the new images_to_act_on algorithm prioritizes selected images over everything else and that would be a problem in culling because here a selection can act like a sub-collection (restricting movement). If we were to always act on all selected images, culling and culling dynamic would not work anymore.

esp. the last sentence : what exactly would not work anymore ? I can only see one : 
Once in "restricted mode" all the images are selected, which means that any action will be applied on the whole set of images... in this case, yes it's a big problem that would void completely actual culling concept. But I wonder how the changes proposed here can solve it...
Let's take the context of the "classic" workflow : filter set to "all except rejected", and culling done by rejecteing "bad" images.
- Do you plan to "unselect" all images as soon as you enter in culling ? Then effectively "restricted" mode is void (and dynamic too). In this case, if we want to keep this modes, the only thing I can think of is to put in memory the initial selection when entering culling, before the unselection, and then to refer to this list and not to the selection one...
- Do you plan to have a "special case" for culling in order to not act on the whole selection ? But if I remember correctly that's not the case...

I really think I miss something...

Solarer commented:
> There's no more "automatic" startup switch

That was unintended, my bad. I will fix that. `x` should still have the automatic behavior whereas `c` is always unrestricted. We do not need an explicit button for restricted since restricted does not make sense when there is no image selected and in that case you can just use `x` to enter restricted mode.

> esp. the last sentence : what exactly would not work anymore ? I can only see one :
Once in "restricted mode" all the images are selected, which means that any action will be applied on the whole set of images... in this case, yes it's a big problem that would void completely actual culling concept. But I wonder how the changes proposed here can solve it...

That is the point and your second bullet point is how I want to solve it:
> Do you plan to have a "special case" for culling in order to not act on the whole selection ? But if I remember correctly that's not the case...

For culling restricted I will ignore selection in `images_to_act_on`.  Therefore I need to split the modes. (This is not included in this PR since we wanted to split this in separate PRs.)

AlicVB commented on src/dtgtk/culling.c:
```diff
@@ -1044,9 +1047,6 @@ void dt_culling_init(dt_culling_t *table, int fallback_offset)
     }
-
-    // we now determine if we limit culling images to the selection
-    table->navigate_inside_selection = (!table->selection_sync && inside);
```
fyi, it's here that we currently "automaticaly" decide the restrict-unrestrict mode...
otherwise the code seems quite clean and not too intrusive...

LebedevRI commented:
This PR should probably be rebased, the diff is, uh,
![image](https://github.com/darktable-org/darktable/assets/88600/34cbc314-35e9-4cdc-b686-3ff7ef8b0e5b)

Solarer commented:
Oh, I didn't mean to push to this repo. So I did not check it.
Yeah, will rebase

Solarer commented:
* Did a rebase of the code with master.
* Fixed the issue of button `x` not behaving as expected.
* Made some comments more verbose to better explain what each section of the code does

Solarer commented:
... forgot to sqash my commits

Solarer commented:
still active and waiting for review from @AlicVB

AlicVB commented:
I've just done another try... and I think there's still some issues : 
1. I've still some rebasing issues. I think that this is due to the fact that the git history before your commit is not clean for some reason. You can solve it by creating a new branch based on current master and adding your commit on top of it...
2. both x and c switch to unrestricted mode
3. if you select 4 images, with the cursor hovering one of the 4 selected images, and use `ctrl-x` you don't seems to enter the right mode... (culling with 2 images is shown). Everything is ok if you don't hover a selected image

I've not looked in the code to find where the issues come from, but I can if you need...

Solarer commented:
Thanks for finding the time. 

1. The rebasing issue is to be expected since other devs are not sitting by idle and pushing new code. I will do a final clean-up and rebase when functionality is ok.

2. `x` still works the same "smart" way as it does on master: If images are selected it enters restricted mode and if no images are selected it goes to unrestricted.
`c` is always unrestricted.
I think this was what we agreed on but I can disable the logic so that `x` is always restricted if that makes more sense to you.
3. Sorry but I was not able to reproduce this. From filemanager I always enter culling-dynamic no matter where I place the mouse. I did notice that I can not move from the other culling modes to culling dynamic via `Ctrl-X`  - it returns me to filemanager layout. I will attempt to fix that but this is not what you mean, right?

AlicVB commented:
> 1. The rebasing issue is to be expected since other devs are not sitting by idle and pushing new code. I will do a final clean-up and rebase when functionality is ok.

I'm far from a git specialist, but I fear the issue here is a little bit different, as I have to solve conflict in files absolutely not touched by your PR. I think the easiest way (at least it's what I'm doing) is to start from a copy of darktable/master and cherry-pick your 2 commits. This way, there's no conflict : it's straightforward !

For point 2 and 3, I've spotted the issue : it's a shortcut conflict with previous settings : if you remove shortcutrc files, no problem anymore (but that point needs to be solved anyway, as we can't ask users to delete their shortcutrc files ;))

Oh, and there's also still a bunch of Gtk-CRITICAL... (but I'm sure you are aware of this)

AlicVB commented on src/libs/tools/lighttable.c:
```diff
@@ -110,6 +116,22 @@ static void _lib_lighttable_update_btn(dt_lib_module_t *self)
+  for(GList *l = children; l; l = g_list_delete_link(l, l))
+  {
+    GtkWidget *w = (GtkWidget *)l->data;
+    gtk_toggle_button_set_active(GTK_TOGGLE_BUTTON(w), (w == active));
```
This give an error : `Gtk-CRITICAL **: 22:15:06.321: gtk_toggle_button_set_active: assertion 'GTK_IS_TOGGLE_BUTTON (toggle_button)' failed` each time you press `xè shortcut for example.
This due to the fact that `d->zoom_box` contains other widgets than just toggle buttons...

AlicVB commented on src/libs/tools/lighttable.c:
```diff
@@ -429,6 +518,23 @@ void gui_init(dt_lib_module_t *self)
+  gtk_widget_set_sensitive(d->zoom_entry, (d->layout != DT_LIGHTTABLE_LAYOUT_CULLING_DYNAMIC && !d->fullpreview));
+  gtk_widget_set_sensitive(d->zoom, (d->layout != DT_LIGHTTABLE_LAYOUT_CULLING_DYNAMIC && !d->fullpreview));
+
+  d->layout_culling_restricted = dtgtk_togglebutton_new(dtgtk_cairo_paint_lt_mode_culling_restricted, 0, NULL);
```
Gives an error `Gtk-CRITICAL **: 22:14:32.005: gtk_widget_set_visible: assertion 'GTK_IS_WIDGET (widget)' failed
` because the definition of the widget comes too late in the code : after the call to `_lib_lighttable_update_btn(self);`

AlicVB reviewed (changes requested):
If that can help you, here's where some gtk criticals comes from...

Solarer commented:
@AlicVB in fact I did not see the GTK errors because I did not look into the console output. Thank you for explaining what the issues were, they are fixed now.

I did a `git rebase` with master. That removes all commits from my branch, fast-forwards the branch to latest master and then reapplies all commits. Faster than cherry picking ;) I hope all git issues are resolved now.

Does your `darktablerc` have an existing key binding for `Ctrl-X` which conflicts? If that is the case, what can I do to fix this?

AlicVB commented:
> Does your `darktablerc` have an existing key binding for `Ctrl-X` which conflicts? If that is the case, what can I do to fix this?

yes, I have : 
```
x=views/lighttable/toggle restricted culling mode
x;ctrl=views/lighttable/toggle culling dynamic mode
```
Which are the default shortcuts... So we'll have for sure to handle them, as the vaste majority of users will face the same issue. @dterrahe may help us on that point, as he's the shortcut guru ;)

dterrahe commented:
There's no safe way for us to switch away from old default key shortcuts, because users may _like_ them that way (and if we change them, they may want to change them back and we'll end up fighting the user).

What happens in somewhat recent master (since #14991) is that if a new default key is specified but all users have an old mapping (the previous default) in their shortcutsrc file, then it will override the new default, which will be listed under "disabled defaults" in the shortcuts dialog. The user can then activate the new default by selecting it and pressing Del (delete the override). You should mention this in the release notes.

Solarer commented:
Wait a moment, there is a misunderstanding. I did not touch the `x` and `ctrl-x` key bindings. At least not from a functional point of view. They do the same thing as before but I might have changed the name of the function - if there is a conflict, it is because of that.

`c` is a new key binding that was not used before (to the best of my knowledge).

So we might have the situation that 
1. Some users created a custom key binding for `c` that I would like to use now. And
2. Some users might have the replaced the defaults `x` and `ctrl-x` with something they like better.

In case of (1) the custom key binding of the user should overrule my binding, correct? I don't see an issue with that 
In case of (2) would I double the binding so that my and the users bindings trigger the same action?

Where does the conflict come from?

dterrahe commented:
3. Everybody has "x" as "toggle culling mode" (because it used to be the default). That will override the new meaning you're trying to assign to it. So after upgrading, both c _and_ x will trigger "toggle culling mode" and the new "toggle restricted" shortcut will show as _disabled_.

Have you tried simulating this upgrade process so you can verify this yourself?

Solarer commented:
No I have not. 
Good idea, I will try that

Solarer commented:
So... I was able to fix the issue with the shortcuts it works now fine when updating from a previous version. From my perspective this PR can be pushed. I will update the first message in this thread to reflect the latest change.

wpferguson commented:
Testing...  Everything appears to be fine so far except use `<` to enter zoom mode.  The shortcut system recognizes the keystroke as `shift+,`.

Solarer commented:
Thanks, I totally missed that. Should work now

Solarer commented:
Please let me squash all commits before merging.

wpferguson commented:
Merging probably wont happen until July at the earliest.  We're in feature freeze right now with 4.8 scheduled to release June 21st.

Solarer commented:
:+1: 
Anyways, if the code is fine and ready for merge without any further changes - please let me know.
I will then squash the commits and begin working on the act-on code based on this PR.

AlicVB commented:
I've done a quick test. Just to be sure, I've spotted 2 changes : 
1. the new "c" shortcut to enter culling in "unrestricted" mode.
2. the automatic mode (shortcut "x") as lost one aspect of the the automatic detection : now it switch between restricted/unrestricted based of the presence of a selection > 1 only. Before, it also check if the mouse hovered the selection or not...

So the question : was that mandatory for the next steps ? If not , I would prefer to keep that point as this...

Another question : full preview works the exact same way as culling for restricted/unrestricted. I've not spotted any change there, but do you plan to do some at some point ?

I'll try to have a look at the code asap, but I can't promise any delay........

Solarer commented:
> I've done a quick test. Just to be sure, I've spotted 2 changes :
> the new "c" shortcut to enter culling in "unrestricted" mode.

Yes, that was present starting with my first PR. All other view modes have a key binding and since **c**ulling starts with `c` and the keybinding is not in use, i picked that.

> the automatic mode (shortcut "x") as lost one aspect of the the automatic detection : now it switch between restricted/unrestricted based of the presence of a selection > 1 only. Before, it also check if the mouse hovered the selection or not...
So the question : was that mandatory for the next steps ? If not , I would prefer to keep that point as this...

Err... honestly, I never noticed that before and I worked on this topic for 2 years... Can we just keep it the way it is now? It is just more straight forward. If someone selects images and then wants to enter unrestricted culling, they can now press `c` and it is done. 

> Another question : full preview works the exact same way as culling for restricted/unrestricted. I've not spotted any change there, but do you plan to do some at some point ?

Not planned. Act-on will continue to work only on the 1 full screen image. Since the filmstrip is hidden by default, nobody should be surprised that you interact with the one image the screen is showing you.

AlicVB commented:
> Err... honestly, I never noticed that before and I worked on this topic for 2 years... Can we just keep it the way it is now? It is just more straight forward. If someone selects images and then wants to enter unrestricted culling, they can now press `c` and it is done.

As soon as it's still not the case for full preview, I'm not sure it's really a critical issue.

Somme more remarks : 
- Unlike "x", "c" doesn't allow to exit culling mode. That has been done like that for "x" in order to facilitate the discovering of how to exit culling mode for beginners. It's also like that for "f" btw.
- If you have a selection (with enough images) and you hover an image outside. Pressing "x" start in "restricted" mode (right, see last comments). But the image (in the selection) where culling start depends of which image is hovered. Esp. if one hovered an image after the selection, culling start at the last images of the selection. imho that's strange as we can crawl in selection in normal order
- We have lost the "selection synchronization" in culling mode. Was that mandatory ? Imho that's not a critical feature, but I want to be sure it's not an unwanted side effect of your changes.

Solarer commented:
Pushed a refined version. Also removed a 
`#include "common/collection.h"`
Because my IDE said it is not required. Compiles just fine without it as far as I can tell

AlicVB commented:
> I implemented "c" to act like a switch to activate or deactivate the restricted movement. Since it is not a real view mode (at least it is not listed with the other ones in the UI) it did not feel right when it could be used to return to filemanager layout. I tested both and found this to be more natural.

From an user POV, I think that this can be really confusing, as "X", "F", etc don't work like that. But that's not really critical. Let's see what other think...



> We lost that some 2 years ago but nobody noticed 😅. It was not my doing. What is selected or not does not matter for this PR here - but it will matter for the act on code later. I suggest to revisit the synchronisation when we work on that.

Oops... you're right. That's an indication of an unneeded feature, I guess ;) Maybe we can clean up last bit of related code then...



> If a mouse-over is available, the code tries to start culling there. Since "x" forces the restriction to selected images it snaps to the next selected image. If we are past the final image in the selection it snaps back.
> I did not implement it that way (just default behavior) but I can change it. I guess it does make sense to start at the beginning by default.

iirc, the code you mention is manly used to move to the nearest image when browsing in a "restricted" mode. That point should remain. But if possible, I agree with you that for the "starting" case, it would be better to go to the beginning...

Solarer commented:
Ok, I believe that is all for now. So I will wait for your approval and merge now that 4.8 is out

AlicVB commented:
Well, my 3 comments above are still actual I think...

Esp, it would be great to have some opinions on : 
> From an user POV, I think that this can be really confusing, as "X", "F", etc don't work like that. But that's not really critical. Let's see what other think...

And to fix : 
> iirc, the code you mention is manly used to move to the nearest image when browsing in a "restricted" mode. That point should remain. But if possible, I agree with you that for the "starting" case, it would be better to go to the beginning...

@wpferguson (as you have volunteered to review), @TurboGit, etc

Solarer commented:
@AlicVB  I already implemented that fix in my last commit. It should now start with the first image.

About the `c` button I left my reasoning above. Of course others should test this for themselves
> I implemented "c" to act like a switch to activate or deactivate the restricted movement. Since it is not a real view mode (at least it is not listed with the other ones in the UI) it did not feel right when it could be used to return to filemanager layout. I tested both and found this to be more natural.

AlicVB commented:
> @AlicVB I already implemented that fix in my last commit. It should now start with the first image.

Oh great ! no idea why I have missed that, I may have done the test before your last commit... Sorry !

So there's just the `c` button shortcut then... let's wait for other opinions (otherwise everything ok for me)

wpferguson commented:
Testing...

These are some observations from just trying to use it.

Did someone request the restricted/unrestricted mode?  Toggling in and out of it can be rather disorienting.  

For instance:
* I select 10 images and go into fixed culling mode with 5 images displayed. I go into restricted mode.  I hover over the second image then toggle out of restricted mode and back into it.  The second image becomes the first.  The next time I toggle the images shift again.
* If I go into unrestricted mode, scroll way past the end of the selection, then hit restricted it jumps back to the selection.  I know this is the way it's supposed to work, but I find it disorienting.
* In restricted mode and I change the number of images displayed from 5 to 2.  I exit restricted mode and the display changes back to 5 images displayed.
* If I move the mouse so it's not over any images and then hit a rating button all the images are assigned the rating.  I was surprised by this, thinking that since I wasn't over an image no image would get rated.

My thought process about restricted mode is if I make a selection then go into culling mode, culling should be restricted to the selection.  Why would I go to the effort of making a selection before culling if I didn't want culling restricted to the selection?

AlicVB commented on src/libs/tools/lighttable.c:
```diff
@@ -242,26 +291,65 @@ static void _lib_lighttable_key_accel_toggle_culling_dynamic_mode(dt_action_t *a
+    _lib_lighttable_set_layout(self, d->base_layout);
+  else // otherwise first check, what the selection count is
+  {
+    int sel_count = 0;
+    sqlite3_stmt *stmt;
+    // clang-format off
+    DT_DEBUG_SQLITE3_PREPARE_V2(dt_database_get(darktable.db),
+                                "SELECT count(*) "
+                                "FROM memory.collected_images AS col, main.selected_images as sel "
+                                "WHERE col.imgid=sel.imgid",
+                                -1, &stmt, NULL);
+    // clang-format on
+    if(sqlite3_step(stmt) == SQLITE_ROW) sel_count = sqlite3_column_int(stmt, 0);
+    sqlite3_finalize(stmt);
```
if we replace all that by : `int sel_count = dt_act_on_get_images_nb(TRUE, FALSE);` that will : 
- simplify the code
- avoid current behavior change with current algo
- be future proof once the new algo is done as it will propose another way to compute `dt_act_on_get_images_nb`. esp in the "selection driven algo", it will return the nb of selected images

AlicVB commented on src/libs/tools/lighttable.c:
```diff
@@ -201,7 +227,28 @@ static gboolean _lib_lighttable_layout_btn_release(GtkWidget *w, GdkEventButton
+    {
+      // if selection count > 1 then enter culling restricted
+
+      int sel_count = 0;
+      sqlite3_stmt *stmt;
+      // clang-format off
+      DT_DEBUG_SQLITE3_PREPARE_V2(dt_database_get(darktable.db),
+                                  "SELECT count(*) "
+                                  "FROM memory.collected_images AS col, main.selected_images as sel "
+                                  "WHERE col.imgid=sel.imgid",
+                                  -1, &stmt, NULL);
+      // clang-format on
+      if(sqlite3_step(stmt) == SQLITE_ROW) sel_count = sqlite3_column_int(stmt, 0);
+      sqlite3_finalize(stmt);
```
same here

AlicVB reviewed (changes requested):
After some more test in "real work" with a lot of images, I came to the conclusion that we lost here a quite important feature when using shortcuts : 
Actually, when you have a selection set (let's say 10 selected images of 50), if you hover an image from the selection, you enter culling in restricted mode. If you hover an image outside this selection you enter culling in unrestricted  mode, starting with the image hovered.
That quite handy imho, and this is consistent with the rest of darktable : fullpreview, darkroom, etc...

The proposed changes allow to keep this feature intact for current algo, and is future proof for a new algo (here "selection driven act_on")

AlicVB commented:
> Did someone request the restricted/unrestricted mode?

iiuc, this is needed, because in the context of a "selection driven act_on" algo, as soon as some selection is set, there would be no way to enter culling in unrestricted mode. @Solarer : I'm right ?



> My thought process about restricted mode is if I make a selection then go into culling mode, culling should be restricted to the selection. Why would I go to the effort of making a selection before culling if I didn't want culling restricted to the selection?

I tend to agree with you here...

Maybe we can find a middle way by making the "rectricted" icon just an indicator and not a switch button ? I'm not sure the switch between restricted and unrestricted mode is useful in any workflow (apart to correct an initial mistake of the user, but in this case exit/reentering culling is not so hard...)
That would also "solve" my concern about the `c` shortcut that behave differently from `x` or `f` on second press...

Solarer commented:
@wpferguson and @AlicVB: I am on vacation for 2 weeks without a computer so I cannot test your findings right now. I will come back to it once I am back at home.

> iiuc, this is needed, because in the context of a "selection driven act_on" algo, as soon as some selection is set, there would be no way to enter culling in unrestricted mode. @Solarer : I'm right ?

The main reason for the split is that a selection driven act_on will always prioritize selected images if they exist. In the culling workflow you always have some selected images to restrict your movement but you still want to be able to apply star ratings to INDIVIDUAL images from that selection - if selection is prioritized all would receive the same rating.

So I need to make an exception in the algorithm in case selection in culling is used for movement restriction. The split allows me to check for that and act accordingly.

@AlicVB was referring to the people who use the culling view mode like file manager layout. They enter without selection, move freely around the collection and select images for export and other actions. When they enter another view mode and return, they are stuck in the selection. The button/shortcut gives them a way back to unrestricted movement.

Solarer commented:
The shortcut is not necessary for the act_on code but that is just 2 more lines of code to fix the dead lock I just described. And I am rewriting that code anyways.

Whether it's a toggle or works the same as `x` and `f` does not matter and can be changed. It just looks more like a toggle button in the UI.  So the binding works like one.

Solarer commented:
Back from vacation

@wpferguson 

> I select 10 images and go into fixed culling mode with 5 images displayed. I go into restricted mode. I hover over the second image then toggle out of restricted mode and back into it. The second image becomes the first. The next time I toggle the images shift again.

fixed. (Regarding the disorientation you mentioned I would like to point out that there is no need to press `c` multiple times in a row and that nobody is really meant to do that.)

> In restricted mode and I change the number of images displayed from 5 to 2. I exit restricted mode and the display changes back to 5 images displayed.

fixed

@AlicVB 
> if we replace all that by : int sel_count = dt_act_on_get_images_nb(TRUE, FALSE);

That will not work. We are trying to decide whether we want to enter restricted or unrestricted mode and that decision is not about the number of images that will be acted on but rather whether a selection is present or not. If 1 image is selected, we want to enter culling restricted. If 0 images are selected but 1 image is hovered over by the mouse we want to enter unrestricted. Yet, it both cases `dt_act_on_get_images_nb` would return `1` as the number of images to act on.

AlicVB commented:
> That will not work. We are trying to decide whether we want to enter restricted or unrestricted mode and that decision is not about the number of images that will be acted on but rather whether a selection is present or not. If 1 image is selected, we want to enter culling restricted. If 0 images are selected but 1 image is hovered over by the mouse we want to enter unrestricted. Yet, it both cases `dt_act_on_get_images_nb` would return `1` as the number of images to act on.

When implementing culling (and reworking full preview) there has been a design decision to not consider selection of 1 image as a "real" selection but more as a starting point, so to enter unrestricted mode in this case.
You can see that behavior actually if you select 1 image  and enter culling with toolbar button or hover the selected image and press culling shortcut.

Is there any imperative reason to change that in the context of "selection driven" algo ?

Solarer commented:
Ok, then it should work in case of hovering 1 image.
